File: M04 Lab - Case Study Python APIs.js
const express = require('express');
const app = express();
app.use(express.json());

let books = [];

// Create a new book
app.post('/books', (req, res) => {
    const book = {
        id: Math.floor(Math.random() * 10000), // random id
        book_name: req.body.book_name,
        author: req.body.author,
        publisher: req.body.publisher
    };
    books.push(book);
    res.status(201).send(book);
});

// Get all books
app.get('/books', (req, res) => {
    res.send(books);
});

// Get a book by id
app.get('/books/:id', (req, res) => {
    const book = books.find(b => b.id === parseInt(req.params.id));
    if (!book) return res.status(404).send('The book with the given ID was not found.');
    res.send(book);
});

// Update a book
app.put('/books/:id', (req, res) => {
    const book = books.find(b => b.id === parseInt(req.params.id));
    if (!book) return res.status(404).send('The book with the given ID was not found.');

    book.book_name = req.body.book_name;
    book.author = req.body.author;
    book.publisher = req.body.publisher;

    res.send(book);
});

// Delete a book
app.delete('/books/:id', (req, res) => {
    const book = books.find(b => b.id === parseInt(req.params.id));
    if (!book) return res.status(404).send('The book with the given ID was not found.');

    const index = books.indexOf(book);
    books.splice(index, 1);

    res.send(book);
});

const port = process.env.PORT || 3000;
app.listen(port, () => console.log(`Listening on port ${port}...`));

File: M04 Lab - Case Study Python APIs.py
from flask import Flask, request, jsonify
from flask_sqlalchemy import SQLAlchemy

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:////tmp/test.db'  # Use your own URI
db = SQLAlchemy(app)

class Book(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    book_name = db.Column(db.String(80), nullable=False)
    author = db.Column(db.String(80), nullable=False)
    publisher = db.Column(db.String(80), nullable=False)

@app.route('/books', methods=['POST'])
def create_book():
    new_book_data = request.get_json()
    new_book = Book(**new_book_data)
    db.session.add(new_book)
    db.session.commit()
    return jsonify(new_book_data), 201

@app.route('/books', methods=['GET'])
def get_books():
    books = Book.query.all()
    return jsonify([book.serialize for book in books]), 200

@app.route('/books/<int:id>', methods=['GET'])
def get_book(id):
    book = Book.query.get(id)
    if book is None:
        return jsonify({'error': 'Book not found'}), 404
    return jsonify(book.serialize), 200

@app.route('/books/<int:id>', methods=['PUT'])
def update_book(id):
    updated_book_data = request.get_json()
    book = Book.query.get(id)
    if book is None:
        return jsonify({'error': 'Book not found'}), 404
    book.update(updated_book_data)
    db.session.commit()
    return jsonify(book.serialize), 200

@app.route('/books/<int:id>', methods=['DELETE'])
def delete_book(id):
    book = Book.query.get(id)
    if book is None:
        return jsonify({'error': 'Book not found'}), 404
    db.session.delete(book)
    db.session.commit()
    return jsonify({'result': 'Book deleted'}), 200

if __name__ == '__main__':
    app.run(debug=True)
